dds/DCPS/DiscoveryBase.h

```diff
@@ -191,11 +192,193 @@ namespace OpenDDS {
+                         ACE_TEXT("discovered topic.\n"),
+                         name_.c_str(),
+                         remote.data_type_name_.c_str(),
+                         local_data_type_name_.c_str()));
+            }
+            if (topic_callbacks_) {
+              topic_callbacks_->inconsistent_topic(inconsistent_topic_count_);
+            }
+          } else if (inconsistent_before && !remote.inconsistent_) {
+            --inconsistent_topic_count_;
+            if (DCPS::DCPS_debug_level) {
+              ACE_DEBUG((LM_WARNING,
+                         ACE_TEXT("(%P|%t) EndpointManager::check_inconsistent_topic - WARNING ")
+                         ACE_TEXT("topic %C discovered data type %C now ")
+                         ACE_TEXT("matches known data type %C\n"),
+                         name_.c_str(),
+                         remote.data_type_name_.c_str(),
+                         local_data_type_name_.c_str()));
+            }
+            if (topic_callbacks_) {
+              topic_callbacks_->inconsistent_topic(inconsistent_topic_count_);
+            }
+          }
+
+          if (remote.inconsistent_) {
+            endpoints_.erase(guid);
+          }
+        }
+
+        // Local and remote
+        void remove_pub_sub(DCPS::RepoId guid) {
+          endpoints_.erase(guid);
+
+          RepoId participant_id = guid;
+          participant_id.entityId = ENTITYID_PARTICIPANT;
+
+          typename RemoteTopicMap::iterator remote_topic_iter = remote_topics_.find(participant_id);
+
+          if (remote_topic_iter == remote_topics_.end()) {
+            // It was local.
+            return;
+          }
+
+          RemoteTopic& remote = remote_topic_iter->second;
+          remote.endpoints_.erase(guid);
+
+          if (remote.inconsistent_) {
+            --inconsistent_topic_count_;
+            if (topic_callbacks_) {
+              topic_callbacks_->inconsistent_topic(inconsistent_topic_count_);
+            }
+          }
+
+          if (remote.endpoints_.empty()) {
+            remote_topics_.erase(remote_topic_iter);
+          }
+        }
+
+        const OPENDDS_STRING local_data_type_name() const { return local_data_type_name_; }
+        const DDS::TopicQos local_qos() const { return local_qos_; }
+        DCPS::RepoId topic_id() const { return topic_id_; }
+        bool has_dcps_key() const { return has_dcps_key_; }
+        bool local_is_set() const { return topic_callbacks_; }
+        const RepoIdSet& endpoints() const { return endpoints_; }
+
+        bool is_dead() const {
+          return topic_callbacks_ == 0 && remote_topics_.empty();
+        }
+
+      private:
+        OPENDDS_STRING name_;
+        OPENDDS_STRING local_data_type_name_;
+        DDS::TopicQos local_qos_;
+        DCPS::RepoId topic_id_;
         bool has_dcps_key_;
+        TopicCallbacks* topic_callbacks_;
         RepoIdSet endpoints_;
+
+        struct RemoteTopic {
+          OPENDDS_STRING data_type_name_;
```

Comment:
In normal cases, this is the same as local_data_type_name_.  We could make this an empty string if it matches local.